prottp.go

```diff
@@ -151,39 +152,36 @@ func wrapMethod(service interface{}, m grpc.MethodDesc) http.Handler {
 			}
 			return
 		}
-		writeMessage(isJson, resp.(proto.Message), http.StatusOK, w)
+		WriteMessage(statusCode, resp.(proto.Message), w, r)
 	})
 }
 
-func writeMessage(isJson bool, msg proto.Message, statusCode int, w http.ResponseWriter) {
+// WriteMessage is exported to be used for middleware to return proto message
+func WriteMessage(statusCode int, msg proto.Message, w http.ResponseWriter, r *http.Request) {
 	var err error
+	isJSON := isRequestJSON(r)
+	if isJSON && w.Header().Get("Content-Type") == "" {
```

Comment:
Since you're changing this, how about updating the headers:

## Request

`Content-Type: X`: "I am sending you X"
`Accept: Y`: "Please respond with Y"

## Response:

`Content-Type: X`: "I am sending you X"

---

So when *reading* the request body, look at `Content-Type` to decide if the body is PB or JSON. When *writing* the response body, look at `Accept` to see whether the client wants PB or JSON, and *set* `Content-Type` to the format that the client asked for.

---

Also, if it's simple, it would be good to include the protobuf name in the content type. For JSON:

```
Content-Type: application/json;<package name.message name>
```

and for PB:

```
Content-Type: application/x.prottp;<package name.message name>
```

Reply:
![snip20171019_16](https://user-images.githubusercontent.com/1014/31760097-239bf38c-b4e6-11e7-8512-b23f02bc5aa2.png)

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept

to fully accept accept specification seems quite complicated.

Reply:
> to fully implement accept specification

Ah, sorry, I wasn't really suggesting implementing full support, e.g. q-factor weighting. Just a simple version that can handle either `application/json` or `application/x.prottp`.

Reply:
> to fully implement accept specification seems quite complicated.

It's odd that it's not in the standard library. `Accept-Language` uses the same format...